### PR TITLE
Add network lifecycle management

### PR DESCRIPTION
**Pull request by nickethier**
This PR adds a few notable changes to support lifecycle management of network namespaces.

#### Alloc Runner Hooks
The first big change here was adding a call to initialize platform dependent hooks since the network hook is currently only supported on linux. There is new dependency here on a small packaged names [`ns`](https://github.com/containernetworking/plugins/tree/master/pkg/ns) that only builds on linux due to references to the syscall package that are platform dependent.

The second is adding the actual network hook that includes checking if a driver needs to initialize the network.

#### `plugins/drivers` Changes
New proto messages and rpcs were added to support network namespaces. These aren't implemented by any driver in this PR but I did add a helper struct that can be embedded when driver network management is not needed (aka everything but docker). The typical proto to go conversion funcs are there as well.

### CODE REVIEW

File: client/allocrunner/alloc_runner_linux.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+func netModeToIsolationMode(netMode string) drivers.NetIsolationMode {
+	switch strings.ToLower(netMode) {
+	case "host":
+		return drivers.NetIsolationModeHost
+	case "bridge", "none":
+		return drivers.NetIsolationModeGroup
+	case "driver":
+		return drivers.NetIsolationModeTask
```
Comment (notnoop) [inline client/allocrunner/alloc_runner_linux.go]: I guess these are defined in another PR - but is this a hard-coded list and should we make job resources network mode use task/group/host strings instead?

The reason I'm asking is how this is related to potentially other networking modes, e.g. `macvlan`, `weaves`, etc and how we envision extensibility of network types.

File: client/allocrunner/network_hook_linux.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+		}
+		defer origNS.Close()
+
+		// create a new netns on the current thread
+		err = unix.Unshare(unix.CLONE_NEWNET)
+		if err != nil {
+			return
+		}
+
+		// Put this thread back to the orig ns, since it might get reused (pre go1.10)
```
Comment (notnoop) [inline client/allocrunner/network_hook_linux.go]: curious how you noticed this issue?  I assume you aren't running golang 1.10?

File: client/allocrunner/alloc_runner_linux.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		netIsolationMode := netModeToIsolationMode(taskNetMode)
+		if !caps.HasNetIsolationMode(netIsolationMode) {
+			return nil, fmt.Errorf("task %s does not support %q networking mode", task.Name, taskNetMode)
+		}
+
+		// check if the driver needs to create the network and if a different
+		// driver has already claimed it needs to initiate the network
+		if caps.MustInitiateNetwork {
+			if networkInitiator != "" {
+				return nil, fmt.Errorf("tasks %s and %s want to initiate networking but only one is able", networkInitiator, task.Name)
```
[the remark this answers is not in the document]
Comment (nickethier) [reply client/allocrunner/alloc_runner_linux.go]: Yes thats correct which this logic handles, but I see how this error can be misleading.

File: client/allocrunner/alloc_runner_linux.go (hunk 1 above)
Comment (nickethier) [reply client/allocrunner/alloc_runner_linux.go]: Other networking modes will be supported through CNI. Likely the network name for these will be `cni/<network-name>` and this logic will be updated then.

I'd argue against mixing the task/group/host (network isolation mode) naming as they have a slightly different purpose. The net isolation mode is for the driver's perspective and doesn't carry any meaning about the network. Where is the network mode would have deeper meaning, especially once CNI is supported.

File: client/allocrunner/network_hook_linux.go (hunk 2 above)
Comment (nickethier) [reply client/allocrunner/network_hook_linux.go]: This func is actually borrowed from: https://github.com/containernetworking/plugins/blob/master/pkg/testutils/netns_linux.go#L34

Though the background on this is covered pretty well here if interested: https://github.com/containernetworking/plugins/tree/master/pkg/ns#namespace-switching

File: client/allocrunner/alloc_runner_linux.go (hunk 1 above)
Comment (notnoop) [reply client/allocrunner/alloc_runner_linux.go]: Ah - makes sense now.  all good then.

File: client/allocrunner/network_hook_linux.go (hunk 2 above)
Comment (notnoop) [reply client/allocrunner/network_hook_linux.go]: Awesome - might be nice to include a link; also may need to copy the whole file along with the copyright and license notice on top.

File: scheduler/util.go [hunk 4]
```diff
@@ -414,6 +406,26 @@ func tasksUpdated(jobA, jobB *structs.Job, taskGroup string) bool {
 	return false
 }
 
+func networkUpdated(netA, netB []*structs.NetworkResource) bool {
```
Comment (schmichael) [inline scheduler/util.go]: Huh, I wonder why this doesn't use https://godoc.org/github.com/hashicorp/nomad/nomad/structs#Networks.Equals

**Pull request merged**